python_modules/dagster/dagster/_core/definitions/remote_asset_graph.py

```diff
@@ -36,18 +38,169 @@
+        # It is currently not possible to access the code version for an observation definition if a
+        # materialization definition also exists. This needs to be fixed.
+        return self._priority_node.code_version
+
+    @property
+    def check_keys(self) -> AbstractSet[AssetCheckKey]:
+        return self._check_keys
+
+    @property
+    def execution_set_asset_keys(self) -> AbstractSet[AssetKey]:
+        return {k for k in self.execution_set_asset_and_check_keys if isinstance(k, AssetKey)}
+
+    @property
+    def execution_set_asset_and_check_keys(self) -> AbstractSet[AssetKeyOrCheckKey]:
+        return self._execution_set_keys
+
+    ##### REMOTE-SPECIFIC INTERFACE
+
+    @property
+    def job_names(self) -> Sequence[str]:
+        # It is currently not possible to access the job names for an observation definition if a
+        # materialization definition also exists. This needs to be fixed.
+        return self._priority_node.job_names if self.is_executable else []
+
+    @property
+    def priority_repository_handle(self) -> RepositoryHandle:
+        # This property supports existing behavior but it should be phased out, because it relies on
+        # materialization nodes shadowing observation nodes that would otherwise be exposed.
+        return next(
+            itertools.chain(
+                (repo for repo, node in self._repo_node_pairs if node.is_materializable),
+                (repo for repo, node in self._repo_node_pairs if node.is_observable),
+                (repo for repo, node in self._repo_node_pairs),
+            )
+        )
+
+    @property
+    def repository_handles(self) -> Sequence[RepositoryHandle]:
+        return [repo_handle for repo_handle, _ in self._repo_node_pairs]
+
+    ##### HELPERS
+
+    @cached_property
+    def _priority_node(self) -> "ExternalAssetNode":
+        # Return a materialization node if it exists, otherwise return an observable node if it
+        # exists, otherwise return any node. This exists to preserve implicit behavior, where the
+        # materialization node was previously preferred over the observable node. This is a
+        # temporary measure until we can appropriately scope the accessors that could apply to
+        # either a materialization or observation node.
+        return next(
+            itertools.chain(
+                (node for node in self._external_asset_nodes if node.is_materializable),
+                (node for node in self._external_asset_nodes if node.is_observable),
+                (node for node in self._external_asset_nodes),
+            )
+        )
+
+    @cached_property
+    def _materializable_node(self) -> "ExternalAssetNode":
+        return next(node for node in self._external_asset_nodes if node.is_materializable)
+
+    @cached_property
+    def _observable_node(self) -> "ExternalAssetNode":
+        return next((node for node in self._external_asset_nodes if node.is_observable))
```

Comment:
hmmm shouldn't this be `Optional`? I don't see how this is guaranteed to return a value

Reply:
It's intended to error if no observable node is defined (as is `_materializable_node`). It's a private implementation detail used only when we know something is observable